File: Code/Server/src/main/java/de/steilerdev/myVerein/server/model/DivisionRepository.java
/**
 * Copyright (C) 2014 Frank Steiler <frank@steilerdev.de>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package de.steilerdev.myVerein.server.model;

import org.springframework.data.mongodb.repository.MongoRepository;
import org.springframework.data.mongodb.repository.Query;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface DivisionRepository extends MongoRepository<Division, String> {

    public Division findByName(String name);
    public List<Division> findByAdminUser(User adminUser);

    /**
     * Gathers all division names available.
     * @return All divisions, but only their name field is populated.
     */
    @Query(value="{}", fields="{ 'name' : 1}")
    public List<Division> findAllNames();

    /**
     * Gathers all division names available, that contain the String.
     * @param contains The returned division needs to contain the specified String.
     * @return All divisions, containing the specified String, but only their name field is populated.
     */
    @Query(value="{'_id': { $regex: '.*?0.*', $options: 'i' }}", fields="{ 'name' : 1}")
    public List<Division> findAllNamesContainingString(String contains);

    public List<Division> findByParent(Division parent);
}

File: Code/iOS Client/myVerein/myVerein/DivisionRepository.swift
//
// Copyright (C) 2015 Frank Steiler <frank@steilerdev.de>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

//
//  DivisionRepository.swift
//  This file holds all information related with parsing and retrieving a division.
//

import Foundation
import CoreData
import XCGLogger

class DivisionRepository: CoreDataRepository {
  
  // MARK: - Functions used to query the database
  
  func findDivisionByUserMembershipStatus(userMembershipStatus: UserMembershipStatus) -> [Division]? {
    logger.verbose("Retrieving divisions the user's membership status is \(userMembershipStatus)")
    // Create a new fetch request using the Message entity
    let fetchRequest = NSFetchRequest(entityName: DivisionConstants.ClassName)
    
    let predicate = NSPredicate(format: "\(DivisionConstants.RawFields.UserMembershipStatus) == %@", userMembershipStatus.rawValue)
    fetchRequest.predicate = predicate
    
    // Execute the fetch request, and cast the results to an array of LogItem objects
    return executeListRequest(fetchRequest)
  }
  
  func findAllDivisions() -> [Division]? {
    logger.verbose("Retrieving all divisions")
    // Create a new fetch request using the Message entity
    let fetchRequest = NSFetchRequest(entityName: DivisionConstants.ClassName)
    
    // Execute the fetch request, and cast the results to an array of LogItem objects
    return executeListRequest(fetchRequest)
  }
  
  // MARK: - Population of division
  override func populateObject<T : CoreDataObject>(coreDataObject: T, usingDictionary dictionary: [String : AnyObject]) -> (T?, NSError?) {
    if let division = coreDataObject as? Division,
      name = dictionary[DivisionConstants.RemoteDivision.Name] as? String
    {
      if let adminUserDict = dictionary[DivisionConstants.RemoteDivision.AdminUser] as? [String: AnyObject] {
        logger.debug("Found an administrator for the division")
        let userRepository = UserRepository()
        let (adminUser: User?, error) = userRepository.getOrCreateUsingDictionary(adminUserDict, AndSync: true)
        // TODO: Or better?!
        if error != nil && adminUser == nil {
          logger.warning("Unable to parse admin user for division: \(error!.extendedDescription)")
          return (nil, error)
        } else {
          division.admin = adminUser!
        }
      } else {
        logger.debug("No admin user for division")
      }
      
      division.desc = dictionary[DivisionConstants.RemoteDivision.Description] as? String
      division.name = name
      division.lastSynced = NSDate()
      
      logger.info("Succesfully parsed and populaterd division")
      return ((division as! T), nil)
    } else {
      let error = MVError.createError(.MVServerResponseParseError)
      logger.error("Unable to parse division: \(error.extendedDescription)")
      return (nil, error)
    }
  }
}